include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_v3r1.hpp

```diff
@@ -0,0 +1,932 @@
+        CShuffleBlockTransferClusterLengths_MBlock_MPerBlock_NBlock_NPerBlock,
+        CShuffleBlockTransferScalarPerVector_NPerBlock,
+        BlkGemmPipeSched,
+        BlkGemmPipelineVer,
+        ComputeTypeA,
+        ComputeTypeB>;
+
+    struct Argument : public GridwiseGemm::Argument
+    {
+        Argument(const ADataType* p_a_grid_,
+                 const BDataType* p_b_grid_,
+                 const std::array<const void*, NumDTensor> p_ds_,
+                 CDataType* p_c_grid_,
+                 index_t M_,
+                 index_t N_,
+                 index_t K_,
+                 index_t StrideA_,
+                 index_t StrideB_,
+                 std::array<ck::index_t, NumDTensor> StrideDs_,
+                 index_t StrideC_,
+                 index_t k_batch_,
+                 bool is_reduce_)
+            : GridwiseGemm::Argument(p_a_grid_,
+                                     p_b_grid_,
+                                     p_c_grid_,
+                                     M_,
+                                     N_,
+                                     K_,
+                                     StrideA_,
+                                     StrideB_,
+                                     StrideC_,
+                                     k_batch_,
+                                     is_reduce_),
+              p_ds(p_ds_),
+              StrideDs(StrideDs_)
+        {
+        }
+
+        const std::array<const void*, NumDTensor> p_ds;
+        std::array<ck::index_t, NumDTensor> StrideDs;
+    };
+
+    using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+    using Add         = tensor_operation::element_wise::Add;
+
+    using ReduceAdd               = ck::reduce::Add;
+    using InElementwiseOperation  = PassThrough;
+    using OutElementwiseOperation = Add;
+
+    using DeviceReduceInstance = DeviceReduceThreadWiseMultiD<CDataType,  // InDataType,
+                                                              DsDataType, // DsDatatype
+                                                              float,      // AccDataType,
+                                                              CDataType,  // OutDataType,
+                                                              3,          // Rank
+                                                              1,          // NumReduceDim
+                                                              ReduceAdd,
+                                                              InElementwiseOperation,
+                                                              OutElementwiseOperation,
+                                                              256, // BlockSize_,
+                                                              4,   // MThreadSliceSize_,
+                                                              1,   // KThreadSliceSize_,
+                                                              0,   // InSrcVectorDim_,
```

Comment:
Are you sure this is correct? If you want to reduce `{M, N, K}` ->`{M, N}` then I think you should set this as `2` since I believe you  want to reduce `KBatch` values.

Reply:
The Layout of input data is `{KBatch, M, N}`, stride is `{M*N, N, 1}`, so InSrcVectorDim_ set InSrcVectorDim_ to 0